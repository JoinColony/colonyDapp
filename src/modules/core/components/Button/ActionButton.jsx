/* @flow */

import type { ComponentType } from 'react';

import React, { Component } from 'react';

import { log } from '~utils/debug';
import DefaultButton from '~core/Button';

import promiseListener from '../../../../createPromiseListener';

import type { AsyncFunction } from '../../../../createPromiseListener';

// TODO if this object is sealed, there are unspecified props being used
type Props = {
  submit: string,
  success: string,
  error: string,
  values?: Object | (() => Object | Promise<Object>),
  button?: ComponentType<*>,
};

type State = {|
  loading: boolean,
|};

class ActionButton extends Component<Props, State> {
  asyncFunc: AsyncFunction<Object, void>;

  constructor(props: Props) {
    super(props);
    const { submit, success, error } = props;
    this.asyncFunc = promiseListener.createAsyncFunction({
      start: submit,
      resolve: success,
      reject: error,
    });
  }

  state = {
    loading: false,
  };

  componentWillUnmount() {
    this.asyncFunc.unsubscribe();
  }

  handleClick = async () => {
    try {
      const { values: valuesProp = {} } = this.props;
      const values =
        typeof valuesProp === 'function' ? await valuesProp() : valuesProp;
      this.setState({ loading: true });
      await this.asyncFunc.asyncFunction(values);
      this.setState({ loading: false });
    } catch (error) {
      log(error);
      this.setState({ loading: false });
      // TODO: display error somewhere
    }
  };

  render() {
    const {
      submit,
      success,
      error,
      values,
      button: Button = DefaultButton,
      ...props
    } = this.props;
    const { loading } = this.state;
    return <Button onClick={this.handleClick} loading={loading} {...props} />;
  }
}

export default ActionButton;
