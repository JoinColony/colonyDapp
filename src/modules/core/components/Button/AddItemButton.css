.main {
  display: flex;
  align-items: center;
  padding: 0px 24px 0px 6px;
  height: 44px;
  border: none;
  border-radius: 32px;
  background-color: color-mod(var(--action-secondary) alpha(12%));
  font-weight: var(--weight-normal);
  color: var(--action-secondary);

  &:hover {
    cursor: pointer;
  }
}

.main[disabled] {
  background-color: color-mod(var(--text-disabled) alpha(12%));
  color: var(--text-disabled);
  cursor: default;
}

.main i {
  margin-right: 10px;
}

.main svg {
  fill: var(--action-secondary);
}

.main[disabled] svg {
  fill: var(--text-disabled);
}
