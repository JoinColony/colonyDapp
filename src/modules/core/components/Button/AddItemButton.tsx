import React from 'react';
import { FormattedMessage, MessageDescriptor } from 'react-intl';

import Icon from '../Icon';
import styles from './AddItemButton.css';

interface Props {
  /** A `messageDescriptor` that make up the button's text label */
  text: MessageDescriptor;

  disabled?: boolean;

  handleClick?: () => void;
}

const displayName = 'Button.AddItemButton';

const AddItemButton = ({ text, handleClick, disabled }: Props) => {
  return (
    <button
      className={styles.main}
      onClick={handleClick}
      type="button"
      disabled={disabled}
    >
      <Icon name="circle-plus" title={text} appearance={{ size: 'medium' }} />
      <span>
        <FormattedMessage {...text} />
      </span>
    </button>
  );
};

AddItemButton.displayName = displayName;

export default AddItemButton;
