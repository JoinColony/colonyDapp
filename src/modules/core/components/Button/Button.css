@value disabledBackground: rgb(194, 204, 204);
@value disabledColor: rgb(250, 251, 251);
@value ghostBorder: rgb(194, 204, 204);

/*
 * Base `Button` styles
 * Everything inherits from this
 */

.main {
  composes: field from '~styles/fields.css';
  composes: busySpin from '~styles/animations.css';
  display: inline-block;
  padding: 4px 16px;
  height: 27px;
  position: relative;
  border: 1px solid transparent;
  border-radius: var(--radius-tiny);
  background-color: transparent;
  font-size: var(--size-small);
  font-weight: var(--weight-semi);
  text-align: center;
  white-space: nowrap;
  letter-spacing: 0.25px;
  transition: background-color 50ms ease-in;
  cursor: pointer;
}

.main:active {
  outline: none;
}

.main:hover {
  outline: none;
}

/*
 * Loading styles via `aria-busy`
 */

.main[aria-busy="true"] {
  /* We really want to hide the text on the button for the spinner. Same for icons */
  color: transparent !important;
  pointer-events: none;

  & svg {
    stroke: transparent !important;
    fill: transparent !important;
  }
}

.main[aria-busy="true"]::after {
  display: block;
  margin: -8px 0 0 -8px;
  height: 16px;
  width: 16px;
  position: absolute;
  top: 50%;
  left: 50%;
  border: 2px solid var(--grey-5);
  border-radius: 50%;
  border-right-color: transparent;
  border-top-color: transparent;
  content: '';
}

/*
 * Disabled styles via `disabled` attribute
 */

.main[disabled]:not([aria-busy="true"]) {
  border-color: disabledBackground;
  background-color: disabledBackground;
  color: disabledColor;
  cursor: default;
}

/*
 * Arranging buttons relative to other elements
 */

.main + .main {
  margin-left: 10px;
}

/*
 * Sizing
 *
 * Next set of classes will define the button's size
 *
 * Heads's up: don't compose `.main` in the button sizing classes, otherwise you'll end up with `.main` apply-ed two times on the element.
 */

.sizeSmall {
  padding: 0 10px;
  height: 14px;
  font-size: var(--size-tiny);
  font-weight: var(--weight-normal);
}

.sizeLarge {
  padding: 11px 32px;
  height: 43px;
  font-size: var(--size-normal);
}

/*
 * Themeing
 *
 * Note: For buttons you always have to define a theme!
 * Next set of classes will define our various button themes. This is a quick outline:
 * - primary: buttons with primary background color
 * - secondary: buttons with transparent background and no border
 * - danger: button with danger colored background
 * - ghost: buttons with transparent background and border color
 * - underlined-bold: special button with fat underline
 * - blue: button with transparent background blue semi bold text and no padding
 *
 * Some of them will also have an `inverted` version. This is for dark backgrounds.
 */

/*
 * Primary
 * Buttons with primary background color
 */

.themePrimary {
  composes: main;
  border-color: var(--primary);
  background-color: var(--primary);
  color: var(--colony-white);
}

.themePrimary:hover {
  border-color: color-mod(var(--primary) shade(15%));
  background-color: color-mod(var(--primary) shade(15%));
  color: var(--colony-white);
}

.themePrimary[aria-busy="true"]::after {
  border-color: var(--colony-white);
  border-right-color: transparent;
  border-top-color: transparent;
}

/*
 * Secondary
 * Buttons with transparent background and no border
 */

.themeSecondary {
  composes: main;
  font-weight: var(--weight-normal);
  color: var(--text);
}

.themeSecondary:hover {
  color: color-mod(var(--text) shade(30%));
}

.themeSecondary.colorSchemaInverted {
  color: var(--colony-white);
}

.themeSecondary.colorSchemaInverted:hover {
  color: color-mod(var(--colony-white) shade(30%));
}

/*
 * Danger
 * Button with danger colored background
 */

.themeDanger {
  composes: main;
  border-color: var(--danger);
  background-color: var(--danger);
  color: var(--colony-white);
}

.themeDanger:hover {
  border-color: var(--danger);
  background-color: color-mod(var(--danger) shade(15%));
  color: var(--colony-white);
}

.themeDanger[aria-busy="true"]::after {
  border-color: var(--colony-white);
  border-right-color: transparent;
  border-top-color: transparent;
}

/*
 * Ghost
 * buttons with transparent background and border color
 */

.themeGhost {
  composes: main;
  border-color: ghostBorder;
  color: var(--colony-black);
}

.themeGhost:hover {
  border-color: ghostBorder;
  color: color-mod(var(--colony-black) tint(15%));
}

.themeGhost.colorSchemaInverted {
  border-color: var(--colony-white);
  color: var(--colony-white);
}

.themeGhost.colorSchemaInverted:hover {
  border-color: var(--colony-white);
  color: color-mod(var(--colony-white) shade(30%));
}

.themeGhost.colorSchemaGrey {
  padding: 3px 16px;
  border-color: var(--temp-grey-1);
  border-width: 2px;
  color: var(--text);
}

.themeGhost.colorSchemaGrey:hover {
  border-color: var(--temp-grey-1);
  color: color-mod(var(--text) shade(30%));
}

.themeGhost.colorSchemaGrey.sizeLarge {
  padding: 10px 32px;
}

/*
 * Underlined Bold
 * Special button with fat underline
 */

.themeUnderlinedBold {
  composes: main;
  padding-left: 0;
  padding-right: 0;
  color: var(--colony-black);
}

.themeUnderlinedBold:hover {
  color: var(--colony-black);
}

.themeUnderlinedBold::after {
  height: 6px;
  width: 100%;
  position: absolute;
  bottom: 0;
  left: 0;
  border-radius: 6px;
  background-color: var(--temp-grey-1);
  content: '';
}

.themeUnderlinedBold:hover::after {
  background-color: var(--primary);
}

.themeBlue {
  composes: main;
  padding: 0;
  height: 20px;
  font-weight: var(--weight-semi);
  color: var(--sky-blue);
}

.themeBlue:hover {
  text-decoration: underline;
}

.themeBlue[disabled]:not([aria-busy="true"]) {
  border-color: transparent;
  background-color: transparent;
  color: disabledBackground;
  cursor: default;
}

.themeBlue[disabled]:not([aria-busy="true"]):hover {
  text-decoration: none;
}
