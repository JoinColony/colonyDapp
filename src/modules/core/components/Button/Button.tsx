/* eslint-disable react/button-has-type */

import React, { ReactNode } from 'react';
import {
  IntlShape,
  MessageDescriptor,
  MessageValues,
  injectIntl,
} from 'react-intl';

import { NavLink } from 'react-router-dom';
import { useMainClasses } from '~utils/hooks';
import styles from './Button.css';

const displayName = 'Button';

export interface Appearance {
  theme?:
    | 'primary'
    | 'secondary'
    | 'danger'
    | 'dangerLink'
    | 'ghost'
    | 'underlinedBold'
    | 'blue';
  size?: 'small' | 'medium' | 'large';
}

export interface Props {
  /** Appearance object */
  appearance?: Appearance;

  /** `children` to render (only works if `value` is not set) */
  children?: ReactNode;

  /** Overwriting class name(s). Setting this will overwrite the `appearance` object */
  className?: string;

  /** Setting this to `true` will apply disabled styles via `aria-disabled` (disable interactions) */
  disabled?: boolean;

  /** Pass a ref to the `<button>` element */
  innerRef?: (ref: HTMLElement | null) => void;

  /** Use a link instead of a button. Like ReactRouter's `to` property */
  linkTo?: string;

  /** Setting this to `true` will apply loading styles via `aria-busy` (show a spinner) */
  loading?: boolean;

  /** Standard html title attribute. Can be a string or a `messageDescriptor` */
  title?: MessageDescriptor | string;

  /** Button type (button|submit) */
  type?: 'submit' | 'reset' | 'button';

  /** A string or a `messageDescriptor` that make up the button's text label */
  text?: MessageDescriptor | string;

  /** Values for loading text (react-intl interpolation) */
  textValues?: MessageValues;

  /** @ignore injected by `react-intl` */
  intl: IntlShape;
}

/*
 * The button type rule recently intoduced in the airbnb ruleset has some funky
 * functionality right now.
 *
 * See this for context: https://github.com/yannickcr/eslint-plugin-react/issues/1555
 *
 * But in combination with Flow and react-intl you can't get it to actually work
 * in this component using default props.
 *
 * TODO Enable react/button-has-type rule
 *
 * After a non-workaround has been found for this.
 */
const Button = ({
  appearance = { theme: 'primary' },
  children,
  className,
  disabled = false,
  innerRef,
  intl: { formatMessage },
  linkTo,
  loading = false,
  title,
  text,
  textValues,
  type = 'button',
  ...props
}: Props) => {
  const titleText =
    typeof title == 'string' ? title : title && formatMessage(title);
  const buttonText =
    typeof text == 'string' ? text : text && formatMessage(text, textValues);

  const classNames = useMainClasses(appearance, styles, className);

  if (linkTo) {
    return (
      <NavLink className={classNames} to={linkTo} {...props}>
        {buttonText || children}
      </NavLink>
    );
  }

  return (
    <button
      className={classNames}
      disabled={disabled || loading}
      aria-busy={loading}
      title={titleText}
      type={type}
      ref={innerRef}
      {...props}
    >
      {buttonText || children}
    </button>
  );
};

Button.displayName = displayName;

export default injectIntl(Button);
