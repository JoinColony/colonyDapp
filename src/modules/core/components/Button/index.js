/* @flow */

export { default } from './Button.jsx';
export { default as ActionButton } from './ActionButton.jsx';
export { default as ConfirmButton } from './ConfirmButton.jsx';
export { default as DialogActionButton } from './DialogActionButton.jsx';
