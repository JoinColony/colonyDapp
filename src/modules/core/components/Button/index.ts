export { default } from './Button';
export { default as ActionButton } from './ActionButton';
export { default as ConfirmButton } from './ConfirmButton';
export { default as DialogActionButton } from './DialogActionButton';
export { default as IconButton } from './IconButton';
