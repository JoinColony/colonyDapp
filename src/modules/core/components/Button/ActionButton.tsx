import React, { useState, ReactType } from 'react';
import { MessageDescriptor } from 'react-intl';

import { ActionTransformFnType } from '~utils/actions';
import { log } from '~utils/debug';
import { useAsyncFunction, useMounted } from '~utils/hooks';
import DefaultButton from '~core/Button';
import { Props as DefaultButtonProps } from './Button';

interface Props extends DefaultButtonProps {
  button?: ReactType;
  confirmText?: any;
  error: string;
  onConfirmToggled?: (...args: any[]) => void;
  onSuccess?: (result: any) => void;
  submit: string;
  success: string;
  text?: MessageDescriptor | string;
  transform?: ActionTransformFnType;
  values?: any | (() => any | Promise<any>);
}

const ActionButton = ({
  button,
  error,
  submit,
  success,
  onSuccess,
  // @todo Remove `values` once async transform functions are supported
  values,
  transform,
  ...props
}: Props) => {
  const isMountedRef = useMounted();
  const [loading, setLoading] = useState(false);
  const asyncFunction = useAsyncFunction({ submit, error, success, transform });

  const handleClick = async () => {
    let result;
    setLoading(true);
    try {
      const asyncFuncValues =
        typeof values == 'function' ? await values() : values;
      result = await asyncFunction(asyncFuncValues);
      if (isMountedRef.current) setLoading(false);
    } catch (err) {
      log.verbose(err);
      setLoading(false);

      /**
       * @todo : display error somewhere
       */
      return;
    }
    if (typeof onSuccess == 'function') onSuccess(result);
  };

  const Button = button || DefaultButton;
  return <Button onClick={handleClick} loading={loading} {...props} />;
};

export default ActionButton;
