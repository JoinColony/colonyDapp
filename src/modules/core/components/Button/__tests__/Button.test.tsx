/* eslint-env jest */

import React from 'react';
import toJson from 'enzyme-to-json';

import { mountWithIntl } from '../../../../../__tests__/utils';

import Button from '../Button';

describe('Button component', () => {
  test('Renders initial component', () => {
    const wrapper = mountWithIntl(<Button text="Some Text" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('renders a plain value', () => {
    const wrapper = mountWithIntl(<Button text="Some Text" />);
    expect(wrapper.text()).toBe('Some Text');
  });

  test('renders an intl message', () => {
    const intlMessage = {
      id: 'intl id',
      defaultMessage: 'intl default message',
    };
    const wrapper = mountWithIntl(<Button text={intlMessage} />);
    expect(wrapper.text()).toBe('intl default message');
  });

  test('renders a child element', () => {
    const wrapper = mountWithIntl(
      <Button>
        <p>Child test</p>
      </Button>,
    );
    // https://github.com/airbnb/enzyme/blob/master/docs/guides/migration-from-2-to-3.md#children-now-has-slightly-different-meaning
    expect(
      wrapper
        .children()
        .children()
        .html(),
    ).toBe('<p>Child test</p>');
  });

  test('button is busy when loading', () => {
    const wrapper = mountWithIntl(<Button text="Some Text" loading />);
    expect(wrapper.html()).toContain('aria-busy="true"');
  });

  test('button is disabled', () => {
    const wrapper = mountWithIntl(<Button text="Some Text" disabled />);
    expect(wrapper.prop('disabled')).toBe(true);
  });

  test('button has custom props', () => {
    const wrapper = mountWithIntl(
      <Button text="Some Text" data-test="test additional props" />,
    );
    expect(wrapper.prop('data-test')).toBe('test additional props');
  });
});
