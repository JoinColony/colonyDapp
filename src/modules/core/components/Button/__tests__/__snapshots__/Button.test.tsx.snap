// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Renders initial component 1`] = `
<Button
  text="Some Text"
>
  <button
    aria-busy={false}
    className=""
    disabled={false}
    type="button"
  >
    Some Text
  </button>
</Button>
`;
